vaadin-cdi/src/main/java/com/vaadin/cdi/context/RouteScopedContext.java

```diff
@@ -94,29 +268,58 @@ public boolean isActive() {
     }
 
-    private Class convertToKey(Contextual<?> contextual) {
-        if (!(contextual instanceof Bean)) {
-            if (contextual instanceof PassivationCapable) {
-                final String id = ((PassivationCapable) contextual).getId();
-                contextual = beanManager.getPassivationCapableBean(id);
-            } else {
-                throw new IllegalArgumentException(
-                        contextual.getClass().getName()
-                                + " is not of type " + Bean.class.getName());
-            }
+    private RouteStorageKey convertToKey(Contextual<?> contextual) {
+        Bean<?> bean = getBean(contextual);
+        UI ui = UI.getCurrent();
+        Class<?> owner = getOwner(ui, bean);
+        if (!navigationChainHasOwner(ui, owner)) {
+            throw new IllegalStateException(String.format(
+                    "Route owner '%s' instance is not available in the "
+                            + "active navigation components chain: the scope defined by the bean '%s' doesn't exist.",
+                    owner, bean.getBeanClass()));
         }
-        final Bean<?> bean = (Bean<?>) contextual;
-        return bean.getQualifiers()
-                .stream()
+        return contextManager.getKey(ui, owner);
+    }
+
+    private boolean navigationChainHasOwner(UI ui, Class<?> owner) {
+        NavigationData data = ComponentUtil.getData(ui, NavigationData.class);
+        if (owner.equals(data.getNavigationTarget())) {
+            return true;
+        }
+        return data.getLayouts().stream()
+                .anyMatch(clazz -> clazz.equals(owner));
+    }
+
+    @SuppressWarnings("unchecked")
+    private Class<?> getOwner(UI ui, Bean<?> bean) {
+        return bean.getQualifiers().stream()
                 .filter(annotation -> annotation instanceof RouteScopeOwner)
-                .map(annotation -> (Class) (((RouteScopeOwner) annotation).value()))
-                .findFirst()
-                .orElse(bean.getBeanClass());
+                .map(annotation -> (Class<?>) (((RouteScopeOwner) annotation)
+                        .value()))
+                .findFirst().orElseGet(() -> getCurrentNavigationTarget(ui));
     }
 
+    @SuppressWarnings("rawtypes")
+    private Class getCurrentNavigationTarget(UI ui) {
+        NavigationData data = ComponentUtil.getData(ui, NavigationData.class);
+        return data.getNavigationTarget();
+    }
+
+    private Bean<?> getBean(Contextual<?> contextual) {
+        if (contextual instanceof Bean) {
+            return (Bean<?>) contextual;
+        }
+        if (contextual instanceof PassivationCapable) {
+            String id = ((PassivationCapable) contextual).getId();
+            return beanManager.getPassivationCapableBean(id);
+        } else {
+            throw new IllegalArgumentException(contextual.getClass().getName()
+                    + " is not of type " + Bean.class.getName());
```

Comment:
Just noting that in case this error can occur, I would not understand anything about what I should do to fix it if I have broken code in my application. If users can face this due to their mistake, some hint would be great to help the resolve the issue.

Reply:
This is refactored code and is out of the scoped of this PR
https://github.com/vaadin/cdi/blob/master/vaadin-cdi/src/main/java/com/vaadin/cdi/context/RouteScopedContext.java#L108.

Reply:
This is fine. Would be good to improve it later on if there is a good idea what it could say